gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ReferralRequestMapper.java

```diff
@@ -0,0 +1,230 @@
+import org.hl7.v3.TS;
+
+import uk.nhs.adaptors.pss.translator.utils.DateFormatUtil;
+
+public class ReferralRequestMapper {
+    private static final String META_PROFILE = "https://fhir.nhs.uk/STU3/StructureDefinition/CareConnect-GPC-ReferralRequest-1";
+    private static final String IDENTIFIER_SYSTEM = "https://PSSAdaptor/";
+    private static final String PRIORITY_PREFIX = "Priority: ";
+    private static final String ACTION_DATE_PREFIX = "Action Date: ";
+    private static final String PRACTITIONER_REFERENCE_PREFIX = "Practitioner/";
+    private static final String RESP_PARTY_TYPE_CODE = "RESP";
+    private static final String PRF_TYPE_CODE = "PRF";
+    private static final String PPRF_TYPE_CODE = "PPRF";
+
+    public ReferralRequest mapToReferralRequest(RCMRMT030101UK04EhrComposition ehrComposition,
+        RCMRMT030101UK04RequestStatement requestStatement) {
+        var id = requestStatement.getId().get(0).getRoot();
+        var identifier = getIdentifier(id);
+        var notes = getNotes(requestStatement);
+        var reasonCode = getReasonCode(requestStatement.getCode());
+        var authoredOn = getAuthoredOn(requestStatement.getAvailabilityTime());
+        var recipient = getRecipient(requestStatement.getResponsibleParty());
+        var requester = getRequesterAgentReference(requestStatement.getParticipant(), ehrComposition);
+
+        /**
+         * TODO: Implement future referencing
+         * - subject: references a global patient resource for the transaction
+         * - context: references an encounter resource if it has been generated from the ehrComposition
+         */
+
+        return createRequestStatement(id, identifier, notes, reasonCode, authoredOn, recipient, requester);
+    }
+
+    private Identifier getIdentifier(String id) {
+        Identifier identifier = new Identifier()
+                .setSystem(IDENTIFIER_SYSTEM) // TODO: concatenate source practice org id to URL
+                .setValue(id);
+        return identifier;
+    }
+
+    private CodeableConcept getReasonCode(CD reasonCode) {
+        if (reasonCode != null) {
+            return new CodeableConceptMapper().mapToCodeableConcept(reasonCode);
+        }
+        return null;
+    }
+
+    private Date getAuthoredOn(TS availabilityTime) {
+        if (availabilityTime != null) {
+            return DateFormatUtil.parse(availabilityTime.getValue()).getValue();
+        }
+        return null;
+    }
+
+    private Reference getRecipient(RCMRMT030101UK04ResponsibleParty3 responsibleParty) {
+        if (responsiblePartyAgentRefHasIdValue(responsibleParty)) {
+            return new Reference(PRACTITIONER_REFERENCE_PREFIX
+                + responsibleParty.getAgentRef().getId().getRoot());
+        }
+        return null;
+    }
+
+    private boolean responsiblePartyAgentRefHasIdValue(RCMRMT030101UK04ResponsibleParty3 responsibleParty) {
+        return responsibleParty != null
```

Comment:
I have the same concern as with checking typeCode inside hasTypeCode method, maybe it would be better to check it this way:
```suggestion
        return responsibleParty.getTypeCode().stream().anyMatch(RESP_PARTY_TYPE_CODE::equals)
            && responsibleParty.getAgentRef() != null
            && responsibleParty.getAgentRef().getId() != null;
```

Reply:
Unfortunately I need the `responsibleParty != null` check but i've added in your suggestion